From 9ee1d378f856b287aafb2fdef7c88121414c01ba Mon Sep 17 00:00:00 2001
From: Andrii Cherniavskyi <achernia@qti.qualcomm.com>
Date: Thu, 19 Jan 2023 11:30:58 +0100
Subject: [PATCH 4/5] libcamera: pipeline: vivid: Add support of multiple
 capture devices.

Implement support of more than a single vivid camera. The cameras
names are enumerated according to their bus names,
e.g. "platform:vivid-000", "platform:vivid-001"...

Signed-off-by: Andrii Cherniavskyi <achernia@qti.qualcomm.com>
---
 src/libcamera/pipeline/vivid/vivid.cpp | 27 ++++++++++++++++++--------
 1 file changed, 19 insertions(+), 8 deletions(-)

diff --git a/src/libcamera/pipeline/vivid/vivid.cpp b/src/libcamera/pipeline/vivid/vivid.cpp
index 0340a500..5728b07a 100644
--- a/src/libcamera/pipeline/vivid/vivid.cpp
+++ b/src/libcamera/pipeline/vivid/vivid.cpp
@@ -59,14 +59,13 @@ public:
 
 	~VividCameraData()
 	{
-		delete video_;
 	}
 
-	int init();
+	int init(MediaDevice *media);
 	void bufferReady(FrameBuffer *buffer);
 
 	MediaDevice *media_;
-	V4L2VideoDevice *video_;
+	std::unique_ptr<V4L2VideoDevice> video_;
 	Stream stream_;
 };
 
@@ -327,7 +326,6 @@ int PipelineHandlerVivid::queueRequestDevice(Camera *camera, Request *request)
 bool PipelineHandlerVivid::match(DeviceEnumerator *enumerator)
 {
 	DeviceMatch dm("vivid");
-	dm.add("vivid-000-vid-cap");
 
 	MediaDevice *media = acquireMediaDevice(enumerator, dm);
 	if (!media)
@@ -336,21 +334,34 @@ bool PipelineHandlerVivid::match(DeviceEnumerator *enumerator)
 	std::unique_ptr<VividCameraData> data = std::make_unique<VividCameraData>(this, media);
 
 	/* Locate and open the capture video node. */
-	if (data->init())
+	if (data->init(media))
 		return false;
 
 	/* Create and register the camera. */
 	std::set<Stream *> streams{ &data->stream_ };
-	const std::string id = data->video_->deviceName();
+	const std::string id = data->video_->busName();
 	std::shared_ptr<Camera> camera = Camera::create(std::move(data), id, streams);
 	registerCamera(std::move(camera));
 
 	return true;
 }
 
-int VividCameraData::init()
+int VividCameraData::init(MediaDevice *media)
 {
-	video_ = new V4L2VideoDevice(media_->getEntityByName("vivid-000-vid-cap"));
+
+        /* Locate and initialise the camera data with the default video node. */
+        const std::vector<MediaEntity *> &entities = media->entities();
+        auto entity = std::find_if(entities.begin(), entities.end(),
+                                   [](MediaEntity *e) {
+                                           return 1;
+                                   });
+        if (entity == entities.end()) {
+                LOG(VIVID, Error) << "Could not find a default video device";
+                return -ENODEV;
+        }
+
+        video_ = std::make_unique<V4L2VideoDevice>(*entity);
+
 	if (video_->open())
 		return -ENODEV;
 
-- 
2.49.0

